File: tools/python/reroom_verify/loader.py
"""Bounded, fail-closed loading for immutable verification fixtures."""

from __future__ import annotations

import hashlib
import json
import math
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator, FormatChecker


ABSOLUTE_MAX_FILE_BYTES = 33_554_432
ABSOLUTE_MAX_CASES = 2_048
ABSOLUTE_MAX_DEPTH = 64
ABSOLUTE_MAX_PATH_BYTES = 240


class VerificationFailure(ValueError):
    """A deterministic boundary rejection with a normalized class."""

    def __init__(self, rejection_class: str, message: str = "") -> None:
        super().__init__(message or rejection_class)
        self.rejection_class = rejection_class


def _reject_constant(value: str) -> None:
    raise VerificationFailure("numeric_out_of_range", f"non-finite number: {value}")


def _unique_object(pairs: list[tuple[str, Any]]) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for key, value in pairs:
        if key in result:
            raise VerificationFailure("duplicate_name", f"duplicate JSON name: {key}")
        result[key] = value
    return result


def _contains_surrogate(value: str) -> bool:
    return any(0xD800 <= ord(character) <= 0xDFFF for character in value)


def _validate_tree(value: Any, depth: int, max_depth: int) -> None:
    if depth > max_depth:
        raise VerificationFailure("json_parse", "maximum document depth exceeded")
    if isinstance(value, str):
        if _contains_surrogate(value):
            raise VerificationFailure("invalid_unicode", "lone UTF-16 surrogate")
        return
    if isinstance(value, float) and not math.isfinite(value):
        raise VerificationFailure("numeric_out_of_range", "non-finite JSON number")
    if isinstance(value, dict):
        for key, child in value.items():
            if _contains_surrogate(key):
                raise VerificationFailure("invalid_unicode", "lone UTF-16 surrogate in name")
            _validate_tree(child, depth + 1, max_depth)
    elif isinstance(value, list):
        for child in value:
            _validate_tree(child, depth + 1, max_depth)


def parse_json_bytes(raw: bytes, *, max_depth: int = ABSOLUTE_MAX_DEPTH) -> Any:
    """Parse UTF-8 JSON while rejecting duplicate names and invalid scalar values."""

    try:
        text = raw.decode("utf-8", errors="strict")
    except UnicodeDecodeError as error:
        raise VerificationFailure("invalid_unicode", "input is not strict UTF-8") from error
    try:
        value = json.loads(
            text,
            object_pairs_hook=_unique_object,
            parse_constant=_reject_constant,
        )
    except VerificationFailure:
        raise
    except json.JSONDecodeError as error:
        raise VerificationFailure("json_parse", "invalid JSON") from error
    _validate_tree(value, 0, min(max_depth, ABSOLUTE_MAX_DEPTH))
    return value


def sha256_bytes(raw: bytes) -> str:
    return hashlib.sha256(raw).hexdigest()


def resolve_within(root: Path, relative_path: str, *, max_path_bytes: int) -> Path:
    if not isinstance(relative_path, str) or not relative_path:
        raise VerificationFailure("invalid_path", "path must be a non-empty string")
    if len(relative_path.encode("utf-8")) > min(max_path_bytes, ABSOLUTE_MAX_PATH_BYTES):
        raise VerificationFailure("invalid_path", "path exceeds its byte limit")
    candidate = Path(relative_path)
    if candidate.is_absolute() or "\\" in relative_path or "\x00" in relative_path:
        raise VerificationFailure("invalid_path", "path syntax is unsafe")
    resolved_root = root.resolve(strict=True)
    resolved = (resolved_root / candidate).resolve(strict=True)
    if not resolved.is_relative_to(resolved_root):
        raise VerificationFailure("invalid_path", "path escapes repository root")
    if not resolved.is_file():
        raise VerificationFailure("invalid_path", "path is not a regular file")
    return resolved


def read_bounded(
    root: Path,
    relative_path: str,
    *,
    max_file_bytes: int = ABSOLUTE_MAX_FILE_BYTES,
    max_path_bytes: int = ABSOLUTE_MAX_PATH_BYTES,
) -> bytes:
    path = resolve_within(root, relative_path, max_path_bytes=max_path_bytes)
    limit = min(max_file_bytes, ABSOLUTE_MAX_FILE_BYTES)
    size = path.stat().st_size
    if size > limit:
        raise VerificationFailure("json_parse", "file exceeds its byte limit")
    raw = path.read_bytes()
    if len(raw) != size or len(raw) > limit:
        raise VerificationFailure("json_parse", "file size changed while reading")
    return raw


@dataclass(frozen=True)
class LoadedFixture:
    repo_root: Path
    fixture_root: Path
    manifest_path: Path
    manifest: dict[str, Any]
    manifest_sha256: str
    max_file_bytes: int
    max_path_bytes: int
    max_document_depth: int

    def read_fixture_file(self, relative_path: str) -> bytes:
        return read_bounded(
            self.fixture_root,
            relative_path,
            max_file_bytes=self.max_file_bytes,
            max_path_bytes=self.max_path_bytes,
        )

    def read_repo_file(self, relative_path: str) -> bytes:
        return read_bounded(
            self.repo_root,
            relative_path,
            max_file_bytes=self.max_file_bytes,
            max_path_bytes=self.max_path_bytes,
        )


def _validate_manifest_files(fixture: LoadedFixture) -> None:
    manifest = fixture.manifest
    case_ids = [case["case_id"] for case in manifest["cases"]]
    if len(case_ids) > ABSOLUTE_MAX_CASES or case_ids != sorted(case_ids):
        raise VerificationFailure("semantic_invariant", "cases are not bounded and ordered")
    if len(case_ids) != len(set(case_ids)):
        raise VerificationFailure("semantic_invariant", "duplicate case_id")

    for schema_hash in manifest["schema_hashes"]:
        raw = fixture.read_repo_file(schema_hash["relative_path"])
        if sha256_bytes(raw) != schema_hash["sha256"]:
            raise VerificationFailure("digest_mismatch", "contract schema hash mismatch")
        schema = parse_json_bytes(raw, max_depth=fixture.max_document_depth)
        if schema.get("$id") != schema_hash["schema_id"]:
            raise VerificationFailure("semantic_invariant", "contract schema id mismatch")
        Draft202012Validator.check_schema(schema)

    for case in manifest["cases"]:
        input_raw = fixture.read_fixture_file(case["input"]["relative_path"])
        if len(input_raw) != case["input"]["byte_length"]:
            raise VerificationFailure("digest_mismatch", "fixture input length mismatch")
        if sha256_bytes(input_raw) != case["input"]["sha256"]:
            raise VerificationFailure("digest_mismatch", "fixture input hash mismatch")
        for artifact in case["expected"]["artifacts"]:
            expected_raw = fixture.read_fixture_file(artifact["relative_path"])
            if len(expected_raw) != artifact["byte_length"]:
                raise VerificationFailure("digest_mismatch", "oracle artifact length mismatch")
            if sha256_bytes(expected_raw) != artifact["sha256"]:
                raise VerificationFailure("digest_mismatch", "oracle artifact hash mismatch")


def load_fixture(manifest_path: Path | str, *, repo_root: Path | str) -> LoadedFixture:
    repo = Path(repo_root).resolve(strict=True)
    manifest_file = Path(manifest_path)
    if not manifest_file.is_absolute():
        manifest_file = repo / manifest_file
    manifest_file = manifest_file.resolve(strict=True)
    if not manifest_file.is_relative_to(repo) or not manifest_file.is_file():
        raise VerificationFailure("invalid_path", "manifest is outside the repository")
    raw = manifest_file.read_bytes()
    if len(raw) > ABSOLUTE_MAX_FILE_BYTES:
        raise VerificationFailure("json_parse", "manifest exceeds absolute size bound")
    manifest = parse_json_bytes(raw)
    schema_raw = read_bounded(repo, "fixtures/manifest.schema.json")
    schema = parse_json_bytes(schema_raw)
    Draft202012Validator.check_schema(schema)
    errors = list(
        Draft202012Validator(schema, format_checker=FormatChecker()).iter_errors(manifest)
    )
    if errors:
        raise VerificationFailure("schema_validation", errors[0].message)

    limits = manifest["limits"]
    max_file_bytes = min(limits["max_file_bytes"], ABSOLUTE_MAX_FILE_BYTES)
    max_path_bytes = min(limits["max_path_bytes"], ABSOLUTE_MAX_PATH_BYTES)
    max_document_depth = min(limits["max_document_depth"], ABSOLUTE_MAX_DEPTH)
    if limits["max_cases"] > ABSOLUTE_MAX_CASES:
        raise VerificationFailure("semantic_invariant", "manifest case limit is too large")
    fixture = LoadedFixture(
        repo_root=repo,
        fixture_root=manifest_file.parent,
        manifest_path=manifest_file,
        manifest=manifest,
        manifest_sha256=sha256_bytes(raw),
        max_file_bytes=max_file_bytes,
        max_path_bytes=max_path_bytes,
        max_document_depth=max_document_depth,
    )
    _validate_manifest_files(fixture)
    return fixture

File: packages/contracts/src/loader.mjs
import { createHash } from "node:crypto";
import { readFile, realpath, stat } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

import Ajv2020 from "ajv/dist/2020.js";
import addFormats from "ajv-formats";

export const MAX_DOCUMENT_DEPTH = 64;
export const MAX_FILE_BYTES = 33_554_432;
export const MAX_CASES = 2_048;
export const MAX_PATH_BYTES = 240;

const ARCHIVE_PATH =
  /^(?!\/)(?![A-Za-z]:)(?!.*(?:^|\/)\.{1,2}(?:\/|$))(?!.*\\)[A-Za-z0-9_-][A-Za-z0-9._-]*(?:\/[A-Za-z0-9_-][A-Za-z0-9._-]*)*$/;
const MODULE_REPO_ROOT = fileURLToPath(new URL("../../..", import.meta.url));
const REGISTERED_SCHEMAS = new Map([
  ["CON-001", ["urn:reroom:schema:frame-packet:1", "docs/contracts/frame-packet.schema.json"]],
  ["CON-002", ["urn:reroom:schema:rrcap-manifest:1", "docs/contracts/rrcap-manifest.schema.json"]],
  ["CON-003", ["urn:reroom:schema:scene-state:1", "docs/contracts/scene-state.schema.json"]],
  ["CON-004", ["urn:reroom:schema:edit-artifacts:1", "docs/contracts/edit-artifacts.schema.json"]],
  ["CON-005", ["urn:reroom:schema:transaction:1", "docs/contracts/transaction.schema.json"]],
]);

export class RunnerFailure extends Error {
  constructor(rejectionClass, message) {
    super(message);
    this.name = "RunnerFailure";
    this.rejectionClass = rejectionClass;
  }
}

export function sha256Hex(bytes) {
  return createHash("sha256").update(bytes).digest("hex");
}

export async function readBytesBounded(filePath, maximum = MAX_FILE_BYTES) {
  let metadata;
  try {
    metadata = await stat(filePath);
  } catch (error) {
    throw new RunnerFailure("invalid_path", "referenced file is unavailable", { cause: error });
  }
  if (!metadata.isFile() || metadata.size > maximum) {
    throw new RunnerFailure("invalid_path", "referenced file is not a bounded regular file");
  }
  const bytes = await readFile(filePath);
  if (bytes.length !== metadata.size) {
    throw new RunnerFailure("digest_mismatch", "referenced file changed while being read");
  }
  return bytes;
}

function syntaxFailure(message) {
  return new RunnerFailure("json_parse", message);
}

export function parseJsonBytesStrict(bytes, { maxDepth = MAX_DOCUMENT_DEPTH } = {}) {
  let source;
  try {
    source = new TextDecoder("utf-8", { fatal: true }).decode(bytes);
  } catch (error) {
    throw new RunnerFailure("invalid_unicode", "JSON is not valid UTF-8", { cause: error });
  }
  if (source.charCodeAt(0) === 0xfeff) {
    throw syntaxFailure("JSON must not contain a byte-order mark");
  }

  let index = 0;
  const whitespace = () => {
    while (index < source.length) {
      const code = source.charCodeAt(index);
      if (code !== 0x09 && code !== 0x0a && code !== 0x0d && code !== 0x20) return;
      index += 1;
    }
  };
  const requireDepth = (depth) => {
    if (depth > maxDepth) throw syntaxFailure("JSON nesting exceeds the configured limit");
  };
  const parseHex = () => {
    const digits = source.slice(index, index + 4);
    if (!/^[0-9a-fA-F]{4}$/.test(digits)) throw syntaxFailure("invalid Unicode escape");
    index += 4;
    return Number.parseInt(digits, 16);
  };
  const parseString = () => {
    if (source[index] !== '"') throw syntaxFailure("expected JSON string");
    index += 1;
    let result = "";
    while (index < source.length) {
      const character = source[index++];
      if (character === '"') return result;
      if (character.charCodeAt(0) < 0x20) throw syntaxFailure("unescaped control character");
      if (character !== "\\") {
        const code = character.charCodeAt(0);
        if (code >= 0xd800 && code <= 0xdfff) {
          throw new RunnerFailure("invalid_unicode", "JSON contains an unpaired Unicode surrogate");
        }
        result += character;
        continue;
      }
      if (index >= source.length) throw syntaxFailure("unterminated JSON escape");
      const escapeCharacter = source[index++];
      const simple = {
        '"': '"',
        "\\": "\\",
        "/": "/",
        b: "\b",
        f: "\f",
        n: "\n",
        r: "\r",
        t: "\t",
      };
      if (Object.hasOwn(simple, escapeCharacter)) {
        result += simple[escapeCharacter];
        continue;
      }
      if (escapeCharacter !== "u") throw syntaxFailure("invalid JSON escape");
      const first = parseHex();
      if (first >= 0xdc00 && first <= 0xdfff) {
        throw new RunnerFailure("invalid_unicode", "JSON contains an unpaired low surrogate");
      }
      if (first >= 0xd800 && first <= 0xdbff) {
        if (source.slice(index, index + 2) !== "\\u") {
          throw new RunnerFailure("invalid_unicode", "JSON contains an unpaired high surrogate");
        }
        index += 2;
        const second = parseHex();
        if (second < 0xdc00 || second > 0xdfff) {
          throw new RunnerFailure("invalid_unicode", "JSON contains an invalid surrogate pair");
        }
        result += String.fromCodePoint(0x10000 + ((first - 0xd800) << 10) + second - 0xdc00);
      } else {
        result += String.fromCharCode(first);
      }
    }
    throw syntaxFailure("unterminated JSON string");
  };

  const parseValue = (depth) => {
    requireDepth(depth);
    whitespace();
    const character = source[index];
    if (character === '"') return parseString();
    if (character === "{") {
      index += 1;
      const value = {};
      const keys = new Set();
      whitespace();
      if (source[index] === "}") {
        index += 1;
        return value;
      }
      while (true) {
        whitespace();
        const key = parseString();
        if (keys.has(key))
          throw new RunnerFailure("duplicate_name", "JSON object contains a duplicate member name");
        keys.add(key);
        whitespace();
        if (source[index++] !== ":") throw syntaxFailure("expected ':' after object member name");
        value[key] = parseValue(depth + 1);
        whitespace();
        const delimiter = source[index++];
        if (delimiter === "}") return value;
        if (delimiter !== ",") throw syntaxFailure("expected ',' or '}' in object");
      }
    }
    if (character === "[") {
      index += 1;
      const value = [];
      whitespace();
      if (source[index] === "]") {
        index += 1;
        return value;
      }
      while (true) {
        value.push(parseValue(depth + 1));
        whitespace();
        const delimiter = source[index++];
        if (delimiter === "]") return value;
        if (delimiter !== ",") throw syntaxFailure("expected ',' or ']' in array");
      }
    }
    for (const [literal, value] of [
      ["true", true],
      ["false", false],
      ["null", null],
    ]) {
      if (source.startsWith(literal, index)) {
        index += literal.length;
        return value;
      }
    }
    const match = /^-?(?:0|[1-9][0-9]*)(?:\.[0-9]+)?(?:[eE][+-]?[0-9]+)?/.exec(source.slice(index));
    if (!match) throw syntaxFailure("invalid JSON value");
    index += match[0].length;
    const value = Number(match[0]);
    if (!Number.isFinite(value))
      throw new RunnerFailure("numeric_out_of_range", "JSON number is not finite");
    return value;
  };

  const value = parseValue(1);
  whitespace();
  if (index !== source.length) throw syntaxFailure("trailing data after JSON value");
  return value;
}

export async function readJsonStrict(filePath, options) {
  return parseJsonBytesStrict(await readBytesBounded(filePath, options?.maxBytes), options);
}

export function validateArchivePath(relativePath) {
  if (
    typeof relativePath !== "string" ||
    Buffer.byteLength(relativePath, "utf8") > MAX_PATH_BYTES ||
    !ARCHIVE_PATH.test(relativePath)
  ) {
    throw new RunnerFailure("invalid_path", "unsafe archive-relative path");
  }
}

export async function resolveContained(
  base,
  relativePath,
  { archivePath = true, containmentRoot = base } = {},
) {
  if (archivePath) validateArchivePath(relativePath);
  if (typeof relativePath !== "string" || relativePath.includes("\0")) {
    throw new RunnerFailure("invalid_path", "invalid file reference");
  }
  try {
    const [root, candidate] = await Promise.all([
      realpath(containmentRoot),
      realpath(path.resolve(base, relativePath)),
    ]);
    const relation = path.relative(root, candidate);
    if (
      relation === "" ||
      (!relation.startsWith(`..${path.sep}`) && relation !== ".." && !path.isAbsolute(relation))
    ) {
      return candidate;
    }
  } catch (error) {
    throw new RunnerFailure("invalid_path", "file reference cannot be resolved safely", {
      cause: error,
    });
  }
  throw new RunnerFailure("invalid_path", "file reference escapes its allowed root");
}

export async function readFixtureFile(fixture, reference) {
  const filePath = await resolveContained(fixture.fixtureRoot, reference.relative_path);
  const bytes = await readBytesBounded(
    filePath,
    Math.min(MAX_FILE_BYTES, reference.byte_length + 1),
  );
  if (bytes.length !== reference.byte_length || sha256Hex(bytes) !== reference.sha256) {
    throw new RunnerFailure(
      "digest_mismatch",
      "fixture file does not match its immutable reference",
    );
  }
  return bytes;
}

export async function readFixtureInternal(fixture, relativePath, base = fixture.fixtureRoot) {
  const filePath = await resolveContained(base, relativePath, {
    archivePath: false,
    containmentRoot: fixture.repoRoot,
  });
  return readBytesBounded(filePath);
}

function createAjv() {
  const ajv = new Ajv2020({
    allErrors: true,
    allowUnionTypes: true,
    strictSchema: true,
    strictTypes: false,
    strictTuples: false,
  });
  addFormats(ajv);
  return ajv;
}

export async function loadFixture(manifestPath, { repoRoot } = {}) {
  const absoluteManifest = path.resolve(manifestPath);
  const root = path.resolve(repoRoot ?? MODULE_REPO_ROOT);
  const manifestBytes = await readBytesBounded(absoluteManifest);
  const manifest = parseJsonBytesStrict(manifestBytes);
  const manifestSchema = await readJsonStrict(path.join(root, "fixtures/manifest.schema.json"));
  const validate = createAjv().compile(manifestSchema);
  if (!validate(manifest))
    throw new RunnerFailure(
      "schema_validation",
      "fixture manifest does not satisfy FixtureManifestV1",
    );
  if (manifest.cases.length > MAX_CASES)
    throw new RunnerFailure("schema_validation", "fixture contains too many cases");
  const caseIds = manifest.cases.map(({ case_id: caseId }) => caseId);
  if (
    new Set(caseIds).size !== caseIds.length ||
    caseIds.some((caseId, index) => index > 0 && caseIds[index - 1] >= caseId)
  ) {
    throw new RunnerFailure(
      "semantic_invariant",
      "fixture case IDs must be unique and strictly lexicographic",
    );
  }

  const schemas = new Map();
  for (const reference of manifest.schema_hashes) {
    const registered = REGISTERED_SCHEMAS.get(reference.contract_id);
    if (
      !registered ||
      registered[0] !== reference.schema_id ||
      registered[1] !== reference.relative_path
    ) {
      throw new RunnerFailure(
        "schema_validation",
        "fixture schema registry tuple is not canonical",
      );
    }
    const schemaPath = await resolveContained(root, reference.relative_path);
    const schemaBytes = await readBytesBounded(schemaPath);
    if (sha256Hex(schemaBytes) !== reference.sha256)
      throw new RunnerFailure("digest_mismatch", "frozen schema digest mismatch");
    const schema = parseJsonBytesStrict(schemaBytes);
    if (schema.$id !== reference.schema_id)
      throw new RunnerFailure("schema_validation", "frozen schema ID mismatch");
    schemas.set(reference.contract_id, schema);
  }

  const fixture = {
    repoRoot: root,
    fixtureRoot: path.dirname(absoluteManifest),
    manifestPath: absoluteManifest,
    manifestBytes,
    manifestSha256: sha256Hex(manifestBytes),
    manifest,
    schemas,
  };
  for (const fixtureCase of manifest.cases) {
    await readFixtureFile(fixture, fixtureCase.input);
    for (const artifact of fixtureCase.expected.artifacts) await readFixtureFile(fixture, artifact);
  }
  return fixture;
}

function decodePointer(pointer) {
  if (pointer === "") return [];
  if (typeof pointer !== "string" || !pointer.startsWith("/"))
    throw new RunnerFailure("schema_validation", "invalid mutation pointer");
  return pointer
    .slice(1)
    .split("/")
    .map((part) => part.replaceAll("~1", "/").replaceAll("~0", "~"));
}

export function applyJsonMutations(source, mutations) {
  const value = structuredClone(source);
  for (const mutation of mutations) {
    if (!["add", "replace"].includes(mutation.op))
      throw new RunnerFailure("schema_validation", "unsupported JSON mutation operation");
    const parts = decodePointer(mutation.pointer);
    if (parts.length === 0)
      throw new RunnerFailure("schema_validation", "root mutation is not supported");
    let target = value;
    for (const part of parts.slice(0, -1)) {
      if (target === null || typeof target !== "object" || !(part in target))
        throw new RunnerFailure("schema_validation", "mutation pointer does not resolve");
      target = target[part];
    }
    const key = parts.at(-1);
    if (
      mutation.op === "replace" &&
      (target === null || typeof target !== "object" || !(key in target))
    ) {
      throw new RunnerFailure("schema_validation", "replace mutation target is absent");
    }
    target[key] = structuredClone(mutation.value);
  }
  return value;
}

export function caseResult(caseId, verdict, rejectionClass = null, outputArtifacts = []) {
  return {
    case_id: caseId,
    verdict,
    rejection_class: rejectionClass,
    output_artifacts: outputArtifacts,
  };
}

export function rejectedCaseResult(fixtureCase, error) {
  if (!(error instanceof RunnerFailure)) throw error;
  return caseResult(fixtureCase.case_id, "reject", error.rejectionClass, []);
}
